File: lang/data_structure/linked_list/singly_2.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (ll *LinkedList) traverse() {
	fmt.Print("traverse\n")
	node := ll.head
	for node != nil {
		fmt.Printf("%d -> ", node.data)
		node = node.next
	}
	fmt.Print("\n")
}

func (ll *LinkedList) append(node *Node) {
	if ll.head == nil {
		ll.head = node
		return
	}

	h := ll.head
	for h.next != nil {
		h = h.next
	}
	h.next = node
}

func (ll *LinkedList) pop() {
	if ll.head == nil {
		return
	}

	if ll.head.next == nil {
		ll.head = nil
		return
	}

	c := ll.head
	n := ll.head.next
	for n.next != nil {
		c = n
		n = n.next
	}
	c.next = nil
}

func main() {
	node1 := &Node{1, nil}
	node2 := &Node{2, nil}
	node3 := &Node{3, nil}
	ll := LinkedList{}

	ll.append(node1)
	ll.append(node2)
	ll.append(node3)
	ll.traverse()

	ll.pop()
	ll.pop()
	ll.pop()
	ll.traverse()
}

File: lang/data_structure/linked_list/singly_2.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *next;
};
typedef struct node Node;

typedef struct
{
    Node *head;
} LinkedList;

void traverse(LinkedList ll)
{
    printf("traverse\n");
    Node *head = ll.head;
    while (head != NULL)
    {
        printf("%d ->", head->data);
        head = head->next;
    }
    printf("\n");
}

void append(LinkedList *ll, Node *n)
{
    if (ll->head == NULL)
    {
        ll->head = n;
        return;
    }

    Node *head = (ll->head);
    Node *p = head;
    while (p->next != NULL)
    {
        p = p->next;
    }
    p->next = n;
}

void pop(LinkedList *ll)
{
    if (ll->head->next == NULL)
    {
        ll->head = NULL;
        return;
    }

    Node *currentNode = ll->head;
    Node *nextNode = ll->head->next;
    while (nextNode->next != NULL)
    {
        currentNode = nextNode;
        nextNode = nextNode->next;
    }
    currentNode->next = NULL;
}

int main()
{
    Node node1 = {1, NULL};
    Node node2 = {2, NULL};
    Node node3 = {3, NULL};
    node1.next = &node2;
    node2.next = &node3;
    LinkedList *ll = malloc(sizeof(LinkedList));
    ll->head = &node1;
    traverse(*ll);

    Node node4 = {4, NULL};

    // append with not null node
    append(ll, &node4);
    traverse(*ll);

    // pop with not null node
    pop(ll);
    traverse(*ll);
    pop(ll);
    pop(ll);

    // pop with null node
    pop(ll);
    traverse(*ll);

    // append with null node
    LinkedList *ll2 = malloc(sizeof(LinkedList));
    Node node5 = {5, NULL};
    append(ll2, &node5);
    traverse(*ll2);
}
